cpp/src/aztec/honk/commitment_scheme/ipa/ipa.hpp

```diff
@@ -0,0 +1,266 @@
+ * @tparam Fq: The field corresponding to the elliptic curve
+ * @tparam G1: The elliptic curve group
+ */
+template <typename Fr, typename Fq, typename G1> class InnerProductArgument {
+  public:
+    using element = typename G1::element;
+    using affine_element = typename G1::affine_element;
+    struct IpaProof {
+        std::vector<affine_element> L_vec;
+        std::vector<affine_element> R_vec;
+        Fr a_zero;
+    };
+    // To contain the public inputs for IPA proof
+    // For now we are including the aux_generator and round_challenges in public input. They will be computed by the
+    // prover and the verifier by Fiat-Shamir when the challengeGenerator is defined.
+    struct IpaPubInput {
+        element commitment;
+        Fr challenge_point;
+        Fr evaluation;
+        size_t poly_degree;
+        element aux_generator;            // To be removed
+        std::vector<Fr> round_challenges; // To be removed
+    };
+
+    /**
+     * @brief Commit to a polynomial
+     *
+     * @param polynomial The input polynomial in the coefficient form
+     * @param poly_degree The degree of the polynomial
+     * @param G_vector The common set of generators required to compute the commitment, to be replaced by srs
+     *
+     * @return a group element
+     */
+    static element commit(std::vector<Fr>& polynomial, const size_t poly_degree, std::vector<affine_element>& G_vector)
+    {
+        auto pippenger_runtime_state = barretenberg::scalar_multiplication::pippenger_runtime_state(poly_degree);
+        auto commitment = barretenberg::scalar_multiplication::pippenger_without_endomorphism_basis_points(
+            &polynomial[0], &G_vector[0], poly_degree, pippenger_runtime_state);
+        return commitment;
+    }
+
+    /**
+     * @brief Compute an IpaProof for opening a single polynomial at a single evaluation point
+     *
+     * @param ipa_pub_input Data required to compute the opening proof. See spec for more details
+     * @param polynomial The witness polynomial whose opening proof needs to be computed
+     * @param G_vector the common set of generators, to be replaced by the srs
+     *
+     * @return an IpaProof, containing information required to verify whether the commitment is computed correctly and
+     * the polynomial evaluation is correct in the given challenge point.
+     */
+    static IpaProof ipa_prove(const IpaPubInput& ipa_pub_input,
+                              const std::vector<Fr>& polynomial,
+                              const std::vector<affine_element>& G_vector)
+    {
+        IpaProof proof;
+        auto& challenge_point = ipa_pub_input.challenge_point;
+        ASSERT(challenge_point != 0 && "The challenge point should not be zero");
+        const size_t poly_degree = ipa_pub_input.poly_degree;
+        // To check poly_degree is greater than zero and a power of two
+        ASSERT((poly_degree > 0) && (!(poly_degree & (poly_degree - 1))) &&
```

Comment:
The commitment scheme should be able to accommodate commitments where the degree is not a power of 2

i.e. the commitment scheme increases the size of the `polynomial` vector until it is a power of 2 and fills the new entries with zeroes.

This work can be done in a future branch through and isn't a blocker to this PR.

Reply:
Added a Todo for this. Will fix this in the upcoming PR.